### DISCUSSION TITLE
Pivots fields must be defined in `fields()`

### DISCUSSION THREAD

**Discussion by bkuhl** [laravel/nova-issues#5280] [Q&A] [answered]
- Laravel Version: 9.46.0
- Nova Version: 4.19.6
- PHP Version: 8.1.2
- Database Driver & Version: MySQL 8.0.3
- Operating System and Version: OSx 13.0.1 (22A400)
- Browser type and version: Chrome 108.0.5359.124 (Official Build) (arm64)
- Reproduction Repository: https://github.com/###/###

### Description:

When viewing the detail page for a `Game`, my app shows a `belongsToMany` relationship to an `Interest`.  This loads as expected using `Game::fieldsForDetail()` and shows the `Interest` name as expected.  However, when I try to show pivot data, the pivot field _cannot_ be seen.  While troubleshooting this, I figured out that on the ajax request loading the `belongsToMany` relationship, `ResolvesFields::fieldsMethod()` is indicating `Game::fields()` should be used to pull pivot fields, _not_ `Game::fieldsForData()`.

It's unclear to me if this is a bug or I'm misunderstanding the design, but I think at the very least I should create this issue in case it's helpful to others.

Here's the ajax request that's triggering the `fields()` to be called instead of `fieldsForDetails()`

```
curl 'https://my-local-site.com/nova-api/interests?search=&filters=W10%3D&orderBy=&perPage=5&trashed=&page=1&viaResource=games&viaResourceId=122&viaRelationship=availableInterests&relationshipType=belongsToMany' \
  -H 'authority: my-local-site.com' \
  -H 'accept: application/json, text/plain, */*' \
  -H 'accept-language: en-US,en;q=0.9' \
  -H 'referer: https://my-local-site.com/nova/resources/games/122' \
```

#### My Setup

**Game Resource**
```php
    public function fieldsForDetail(NovaRequest $request)
    {
        return [
            // fields removed
            // I expected this to be sufficient for the `Boolean` field to show up on the Game details page
            // However, I needed to define this exact field in `fields()` and then it showed up
            BelongsToMany::make('Interests', 'availableInterests', Interest::class)
                ->fields(function (NovaRequest $request, $relatedModel) {
                    return [
                        Boolean::make('Unlocked', 'unlocked'),
                    ];
                }),
        ];
    }
```

Related issues: (mainly included these to help others find this issue from other issues)
 - https://github.com/laravel/nova-issues/issues/155
 - https://github.com/laravel/nova-issues/issues/3615

**Comment by crynobone** [comment, marked as answer]
The api endpoint shows without resourceId and this clearly indicate it is fetching `ResourceIndexRequest` and therefore would use `fieldsForIndex()` or `fields()`.

**Comment by bkuhl** [comment]
Notice the referrer URL is a detail page for a resource.  Here's a screenshot of the lock field not being shown:

<img width="961" alt="Screenshot 2023-01-25 at 7 20 00 PM" src="https://user-images.githubusercontent.com/524933/214724538-76f9f6f8-17a4-4469-bc7e-57ac98ea19a4.png">

Perhaps the bug is in VueJS then, initiating the wrong backend request.

**Comment by crynobone** [comment]
> Notice the referrer URL is a detail page for a resource.

That's the parent resource URL, not the BelongsToMany Index request as posted above:

![image](https://user-images.githubusercontent.com/172966/214726330-1e540fc9-79d7-48e1-bcd7-06960ac56e68.png)


> Perhaps the bug is in VueJS then, initiating the wrong backend request.

No.

**Comment by bkuhl** [comment]
Ohh, ok.  I just reproduced it again... I must've had my tabs mixed up or something when I first submitted this and I was focused on the `Network` tab in chrome and thus provided the wrong curl as you mentioned.  Here's the ajax request from `https://my-local-site.com/nova/resources/games/122` that tries:

```
curl 'https://local.mysite-env.com/nova-api/interests?search=&filters=W10%3D&orderBy=&perPage=5&trashed=&page=1&viaResource=games&viaResourceId=121&viaRelationship=availableInterests&relationshipType=belongsToMany' \
  -H 'authority: my-local-site.com' \
  -H 'referer: https://my-local-site.com/nova/resources/games/121'
```


I added a `var_dump(get_class($request));` on `ResolvesFields::fieldsMethod()` first thing and saw this output for the above request:

<img width="446" alt="Screenshot 2023-01-25 at 7 45 01 PM" src="https://user-images.githubusercontent.com/524933/214728944-01bc2b01-5fa0-4b7c-8fdf-507d70d17f3e.png">

**Comment by bkuhl** [comment]
@crynobone Can you help me understand why you converted this to a discussion?  Was my last message some further indication this isn't a bug but I'm doing something wrong?

**Comment by mathieufrh** [comment]
This behaviour is really strange and according to me, this is a real bug.

I am using fields, fieldsForIndex and fieldsForDetails. When defining a BelongsToMany with a fields method to add pivot fields, it is ONLY working if I add it to fields, fieldsForIndex and fieldsForDetails. I understand that showing many related resources of type A from the details page of a resource B is using the index view of A, but why requiring A to exists in the fieldsForIndex of resource B ? It does not really make sense.
